keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,617 @@
+endpoints they should route to.
+
+For example, for a Service with 3 endpoints, the EndpointSlice controller may
+create an EndpointSlice with endpoints that look like this:
+
+```
+- addresses: ["10.1.2.3"]
+  zone: "zone-a"
+  hints:
+    zone: "zone-a"
+- addresses: ["10.1.2.3"]
+  zone: "zone-b"
+  hints:
+    zone: "zone-b"
+- addresses: ["10.1.2.3"]
+  zone: "zone-a"
+  hints:
+    zone: "zone-c"
+```
+
+In the above example, 2 endpoints are in zone-a and 1 endpoint is in zone-b. The
+hints help ensure that each zone will have a single endpoint to consume by
+adding a hint to the third endpoint that it should be consumed by "zone-c".
+
+This functionality will be enabled by a `TopologyAwareHints` feature gate along
+with a Service annotation.
+
+### Risks and Mitigations
+
+- In a scenario where all traffic originates from a single zone there is a
+  chance that endpoints in that zone will be overloaded while endpoints in other
+  zones receive little to no traffic. Without some sort of feedback (out of
+  scope) this will not self-rectify.
+- Autoscaling will not behave well if only a single zone is receiving large
+  amounts of traffic. This could potentially be mitigated by separating
+  deployments and HPAs per zone.
+- Services with ExternalTrafficPolicy=local will need special treatment here.
+  This approach could result in a situation where an endpoint on a Node is
+  delivered to a separate underprovisioned zone. The simplest approach would be
+  to disable this functionality altogether.
+- When this feature is transitioning between enabled and disabled states, there
+  will be a brief point in time where only some EndpointSlices have hints. That
+  could temporarily result in traffic being routed to a small subset of
+  endpoints. To avoid this, we only filter out endpoints that have a hint set
+  to a different zone. If a hint is not set for an endpoint, it will be included
+  by all instances of kube-proxy.
+
+## Design Details
+
+### Assumptions
+
+- Incoming traffic is proportional to the number of allocatable CPU cores in a
+  zone. Although this is an imperfect metric, it is the best available way of
+  predicting how much traffic will be received in a zone. If we are unable to
+  derive the number of allocatable cores in a zone we will fall back to the
+  number of nodes in that zone.
+- Service capacity is proportional to the number of endpoints in a zone. This
+  assumes that each endpoint has equivalent capacity. Although this is not
+  always true, it usually is. We can explore ways to deal with variable capacity
+  endpoints in the future.
+
+### Identifying Zones
+
+The EndpointSlice controller reads the standard `topology.kubernetes.io/zone`
+annotation on Nodes to determine which zone a Pod is running in. Kube-Proxy
+would be updated to read the same information to identify which zone it is
+running in.
+
+### Configuration
+
+A new `service.kubernetes.io/topology-aware-routing` annotation will be
```

Comment:
Instead of an annotation, I wonder if `TopologyAware` (or just `Topology`) can one of the possible values for `internalTrafficPolicy`. That way we can validate and reject it when `externalTrafficPolicy: Local` or when `internalTrafficPolicy: Local|PreferLocal`.

Reply:
I guess we can also validate against annotation but not sure what the precedent for that is

Reply:
Good call, based on the suggestion here and in that KEP (https://github.com/kubernetes/enhancements/pull/2441#discussion_r572374421) and some follow up discussion in Slack, I think it makes sense to integrate this with the `TrafficPolicy` field instead of adding a new annotation. I've updated this KEP to reflect that. I think you provided a good summary in Slack of what we're thinking here:

- if `externalTrafficPolicy=Cluster`, fall back to `trafficPolicy` for external sources
- if `externalTrafficPolicy=Local` -- this rule only takes precedent for external sources, with internal traffic following `trafficPolicy`